src/components/data-loader.tsx

```diff
@@ -8,30 +14,65 @@ import '../styles/components/data-loader.scss';
 
 const ioSupport = 'IntersectionObserver' in window;
 
-const withDataLoader = (BaseComponent) => {
-  const Wrapper = (props) => {
-    const {
-      onLoadMoreItems,
-      hasMoreData,
-      data,
-      loaderComponent,
-      clickToLoad,
-    } = props;
+type WrapperProps = {
+  /**
+   * Callback to request more items if user scrolled to the bottom of the scroll-container or if
+   * the scroll-container isn't scrollable yet because not enough items have been loaded yet. If
+   * not provided this component will simply pass the data prop to the BaseComponent to be rendered
+   * without observing scroll or triggering more data loading.
+   */
+  onLoadMoreItems: () => void;
+  /**
+   * A boolean to indicate that the parent has more items to provide.
+   */
+  hasMoreData: boolean;
+  /**
+   * A custom loader component
+   */
+  // eslint-disable-next-line react/require-default-props
+  loaderComponent?: ReactNode;
+  /**
+   * Data that is being represented in the wrapped component
+   */
+  data: unknown[];
+  /**
+   * Use a button to load more data instead of having infinite scrolling
+   */
+  // eslint-disable-next-line react/require-default-props
+  clickToLoad?: boolean;
+};
+
+const withDataLoader = (
+  BaseComponent: (props: { data: unknown[] }) => JSX.Element
+) => {
+  const Wrapper = ({
```

Comment:
And 
```suggestion
  const Wrapper: FC<WrapperProps> = ({
```
As FC types children for us, and in a slightly more generic way using `ReactNode`rather than using `JSX.Element`